docs/preview/features/message-pumps/service-bus-azure-functions.md

```diff
@@ -0,0 +1,150 @@
+
+Processing shipment messages:
+
+```csharp
+public class ShipmentServiceBusMessageHandler : IAzureServiceBusMessageHandler<Shipment>
+{
+    private readonly ILogger _logger;
+
+    public ShipmentServiceBusMessageHandler(ILogger<ShipmentServiceBusMessageHandler> logger)
+    {
+        _logger = logger;
+    }
+
+    public async Task ProcessMessageAsync(
+        Shipment shipment,
+        AzureServiceBusMessageContext messageContext,
+        MessageCorrelationInfo correlationInfo,
+        CancellationToken cancellationToken)
+    {
+        _logger.LogInformation("Processing shipment {ShipmentId} for order #{OrderId}", shipment.Id, shipment.Order.Id);
+    }
+}
+```
+
+Processing order messages:
+
+```csharp
+public class OrderServiceBusMessageHandler : IAzureServiceBusMessageHandler<Order>
+{
+    private readonly ILogger _logger;
+
+    public OrderServiceBusMessageHandler(ILogger<OrderServiceBusMessageHandler> logger)
+    {
+        _logger = logger;
+    }
+
+    public async Task ProcessMessageAsync(
+        Order order,
+        AzureServiceBusMessageContext messageContext,
+        MessageCorrelationInfo correlationInfo,
+        CancellationToken cancellationToken)
+    {
+        log.LogInformation("Processing order {OrderId} for {OrderAmount} units of {OrderArticle} bought by {CustomerFirstName} {CustomerLastName}", order.Id, order.Amount, order.ArticleNumber, order.Customer.
+    }
+}
+```
+
+Registering both message handlers:
+
+```csharp
+[assembly: FunctionsStartup(typeof(Startup))]
+namespace MessageProcessing
+{
+    public class Startup : FunctionsStartup
+    {
+        public override void Configure(IFunctionsHostBuilder builder)
+        {
+            builder.Services.WithServiceBusMessageHandler<OrderServiceBusMessageHandler, Order>(messageContext => messageContext.UserProperties["MessageType"] == MessageType.Order)
+                            .WithServiceBusMessageHandler<ShipmentServiceBusMessageHandler, Shipment>(messageContext => messageContext.UserProperties["MessageType"] == MessageType.Shipment);
+        }
+    }
+}
+```
+
+## Activating message routing
+To actually use these registered message handles in the Azure Function, you'll have to register the Azure Service Bus message router.
+This can easily done with the `.AddServiceBusRouting` extension:
+
+```csharp
+[assembly: FunctionsStartup(typeof(Startup))]
+namespace MessageProcessing
+{
+    public class Startup : FunctionsStartup
+    {
+        public override void Configure(IFunctionsHostBuilder builder)
+        {
+            builder.Services.AddServiceBusMessageRouting()
+                            .WithServiceBusMessageHandler<OrderServiceBusMessageHandler, Order>(messageContext => messageContext.UserProperties["MessageType"] == MessageType.Order)
+                            .WithServiceBusMessageHandler<ShipmentServiceBusMessageHandler, Shipment>(messageContext => messageContext.UserProperties["MessageType"] == MessageType.Shipment);
```

Comment:
We could consider to extract these extensions to the seperate 'core' project of Service Bus and create some new extensions for the specific Azure Functions `IFunctionHostBuilder` type. But not sure if this is 
1. worth the effort, since it will be first of all require some work to duplicate all possible combinations to this new type
2. all namespaces will be different then. Message handler interfaces are now in the `...Pumps.ServiceBus` namespace while they will be moved to `...ServiceBus.Core`. We could of course let them exist in this 'old' namespace so we don't break anything; and move it in a later stage.

Reply:
Aha, there's also option 3. where we create a separate extension for message routing for `IFunctionHostBuilder` but return an `IServiceCollection` instead so we can reuse the other message handler extensionns.

Reply:
This last option is maybe the best, because otherwise we would have to include the message pump library in a Azure Functions project, while we only would need the message handlers in that case.

So, proposal to move the message handlers to a the ServiceBus Core project so we can reuse it in both the message pump project and Azure Functions project. The problem I see with this approach, is that the interface for the message handler also would have to move to the Core project and that would be a breaking change.

Reply:
Since we have a specific package already I think we should only provide extensions that use IFunctionHostBuilder as input and call the others under the hood, to be honest; or even introduce a service specific Azure function package (my pref).

Better do it right now than think it would have been better later on :D

Reply:
Ok, hmm, then we should add 64+ extensions which are almost exact duplicates, so we can use this interface instead.

Reply:
The question here was also whether or not to move the message handling interfaces as they are both used in message pumps as Azure Functions projects.

Reply:
> Ok, hmm, then we should add 64+ extensions which are almost exact duplicates, so we can use this interface instead.

Definitely in another issue I would say, if we go through with this.

Reply:
> Ok, hmm, then we should add 64+ extensions which are almost exact duplicates, so we can use this interface instead.

It's ok to leave out for now then, it just makes it less discoverable :)

> The question here was also whether or not to move the message handling interfaces as they are both used in message pumps as Azure Functions projects.

Sharing is ok for me :)

Reply:
Added an issue for this: #154 . Resolving.